CRM/Contribute/BAO/Contribution/Utils.php

```diff
@@ -170,7 +170,19 @@ public static function processConfirm(
             // and always calling it first.
             $form->postProcessHook();
           }
-          $result = $payment->doPayment($paymentParams);
+          $processorParams = $paymentParams;
+          $result = $payment->doPayment($processorParams);
+          if (empty($result['fee_amount']) && is_array($processorParams) && !empty($processorParams['fee_amount'])) {
+            // It's possible a processor might be setting fee_amount in the processorParams
+            // & expecting it to be parsed as this may have happened historically (the array used to be
+            // passed on to weird & wonderful places.)
+            // We don't have to worry about the other return params we support as we can see
+            // below that trxn_id was always taking from what was returned, ditto payment_status_id.
+            // Note we add the is_array check as most processors are not using property bags & we can reasonably
+            // communicate the need to add this to return properties to the very few that are.
+            CRM_Core_Error::deprecatedFunctionWarning('fee amount should be returned from doPayment if it needs to be set');
+            $result['fee_amount'] = $processorParams['fee_amount'];
+          }
```

Comment:
`fee_amount` is pretty badly handled in a lot of cases (ie. not properly recorded). I think we're better not adding any special handling and just fixing the issues recording fee_amount as they come up. We should just expect it to be returned in the usual way and record it on the payment if it is returned. See. eg. https://lab.civicrm.org/extensions/stripe/-/issues/207 though note that I don't think this is specific only to webform+transact API. I just don't have any more examples to hand :-)

Reply:
@mattwire the usual way being via the return array & not being noisy if they get it wrong?